### PR TITLE
Update current flank flags

### PR DESCRIPTION
**Pull request by pawelpasterz**
New flags added
* runTimeout
* ignoreFailedTests
* disableSharding
* smartFlankDisableUpload
* testRunnerClass
* localResultsDir
* numUniformShards
* clientDetails
* testTargetsAlwaysRun
* otherFiles
* networkProfile

### CODE REVIEW

**Comment by pawelpasterz** [comment]
Hey @runningcode , please take a look and let me know what do you think. Thanks!

File: README.md [hunk 1]
```diff
@@ -184,6 +203,38 @@ Keeps the full path of downloaded files from a Google Cloud Storage bucket. Requ
 ### resultsDir
 The name of a unique Google Cloud Storage object within the results bucket where raw test results will be stored. The default is a timestamp with a random suffix.
 
+### runTimeout
+The max time this test run can execute before it is cancelled. s (seconds), m (minutes), h (hours) suffixes are acceptable (default: unlimited).
```
Comment (runningcode) [inline README.md:207]: can you add a little example here? is it like `1h2m2s` or `1h 2m 2s`? or just `1h`?

File: README.md [hunk 2]
```diff
@@ -184,6 +203,38 @@ Keeps the full path of downloaded files from a Google Cloud Storage bucket. Requ
...
+
+### ignoreFailedTests
+Terminate with exit code 0 when there are failed tests. Useful for Fladle and other gradle plugins that don't expect the process to have a non-zero exit code. The JUnit XML is used to determine failure. (default: false)
```
Comment (runningcode) [inline README.md:210]:
```suggestion
Always return successful task completion even when there are failed tests. Useful when parsing JUnit XML to determine failure. (default: false)
```

File: README.md [hunk 3]
```diff
@@ -184,6 +203,38 @@ Keeps the full path of downloaded files from a Google Cloud Storage bucket. Requ
...
+
+### disableSharding
+Disables sharding. Useful for parameterized tests. (default: false)
+
+### smartFlankDisableUpload
+Disables smart flank JUnit XML uploading. Useful for preventing timing data from being updated. (default: false)
```
Comment (runningcode) [inline README.md:216]: What is "smart flank"? Can we link to docs in flank here?

File: README.md [hunk 4]
```diff
@@ -184,6 +203,38 @@ Keeps the full path of downloaded files from a Google Cloud Storage bucket. Requ
...
 The name of a unique Google Cloud Storage object within the results bucket where raw test results will be stored. The default is a timestamp with a random suffix.
 
+### runTimeout
+The max time this test run can execute before it is cancelled. s (seconds), m (minutes), h (hours) suffixes are acceptable (default: unlimited).
+
+### ignoreFailedTests
+Terminate with exit code 0 when there are failed tests. Useful for Fladle and other gradle plugins that don't expect the process to have a non-zero exit code. The JUnit XML is used to determine failure. (default: false)
+
+### disableSharding
+Disables sharding. Useful for parameterized tests. (default: false)
```
Comment (runningcode) [inline README.md:213]: Is this statement correct?
```suggestion
Disables sharding. All tests will run on the same device. Useful for parameterized tests which do not support sharding. (default: false)
```

File: README.md [hunk 5]
```diff
@@ -184,6 +203,38 @@ Keeps the full path of downloaded files from a Google Cloud Storage bucket. Requ
...
+
+### smartFlankDisableUpload
+Disables smart flank JUnit XML uploading. Useful for preventing timing data from being updated. (default: false)
+
+### testRunnerClass
+The fully-qualified Java class name of the instrumentation test runner (default: the last name extracted from the APK manifest).
```
Comment (runningcode) [inline README.md:219]: Is this correct?
```suggestion
The fully-qualified Java class name of the instrumentation test runner (default: the test manifest is parsed to determine the class name).
```

File: README.md [hunk 6]
```diff
@@ -184,6 +203,38 @@ Keeps the full path of downloaded files from a Google Cloud Storage bucket. Requ
...
+
+### localResultsDir
+Local folder to store the test result. Folder is DELETED before each run to ensure only artifacts from the new run are saved.
```
Comment (runningcode) [inline README.md:222]:
```suggestion
Local directory to store the test results. Folder is DELETED before each run to ensure only artifacts from the new run are saved.
```

File: README.md [hunk 7]
```diff
@@ -184,6 +203,38 @@ Keeps the full path of downloaded files from a Google Cloud Storage bucket. Requ
...
+
+### numUniformShards
+Specifies the number of shards into which you want to evenly distribute test cases. The shards are run in parallel on separate devices. For example, if your test execution contains 20 test cases and you specify four shards, each shard executes five test cases. The number of shards should be less than the total number of test cases. The number of shards specified must be >= 1 and <= 50. This option cannot be used along max-test-shards and is not compatible with smart sharding. If you want to take benefits of smart sharding use max-test-shards instead. (default: null)
+
+### clientDetails
+A key-value map of additional details to attach to the test matrix. Arbitrary key-value pairs may be attached to a test matrix to provide additional context about the tests being run. When consuming the test results, such as in Cloud Functions or a CI system, these details can add additional context such as a link to the corresponding pull request.
+
+### testTargetsAlwaysRun
+Always run - these tests are inserted at the beginning of every shard useful if you need to grant permissions or login before other tests run
```
Comment (runningcode) [inline README.md:231]:
```suggestion
Always run - these tests are inserted at the beginning of every shard. Useful if you need to grant permissions or login before other tests run
```

File: README.md [hunk 8]
```diff
@@ -184,6 +203,38 @@ Keeps the full path of downloaded files from a Google Cloud Storage bucket. Requ
...
+
+### localResultsDir
+Local folder to store the test result. Folder is DELETED before each run to ensure only artifacts from the new run are saved.
+
+### numUniformShards
+Specifies the number of shards into which you want to evenly distribute test cases. The shards are run in parallel on separate devices. For example, if your test execution contains 20 test cases and you specify four shards, each shard executes five test cases. The number of shards should be less than the total number of test cases. The number of shards specified must be >= 1 and <= 50. This option cannot be used along max-test-shards and is not compatible with smart sharding. If you want to take benefits of smart sharding use max-test-shards instead. (default: null)
+
+### clientDetails
+A key-value map of additional details to attach to the test matrix. Arbitrary key-value pairs may be attached to a test matrix to provide additional context about the tests being run. When consuming the test results, such as in Cloud Functions or a CI system, these details can add additional context such as a link to the corresponding pull request.
```
Comment (runningcode) [inline README.md:228]: Can we link to gcloud docs here to better understand how these are passed to the device?
Also an example of how to consume these client details? Are these to be consumed on the test device or in scripts?

File: README.md [hunk 9]
```diff
@@ -184,6 +203,38 @@ Keeps the full path of downloaded files from a Google Cloud Storage bucket. Requ
...
+Disables smart flank JUnit XML uploading. Useful for preventing timing data from being updated. (default: false)
+
+### testRunnerClass
+The fully-qualified Java class name of the instrumentation test runner (default: the last name extracted from the APK manifest).
+
+### localResultsDir
+Local folder to store the test result. Folder is DELETED before each run to ensure only artifacts from the new run are saved.
+
+### numUniformShards
+Specifies the number of shards into which you want to evenly distribute test cases. The shards are run in parallel on separate devices. For example, if your test execution contains 20 test cases and you specify four shards, each shard executes five test cases. The number of shards should be less than the total number of test cases. The number of shards specified must be >= 1 and <= 50. This option cannot be used along max-test-shards and is not compatible with smart sharding. If you want to take benefits of smart sharding use max-test-shards instead. (default: null)
```
Comment (runningcode) [inline README.md:225]: Looks good. Can we link to docs on "smart sharding"?

File: README.md [hunk 10]
```diff
@@ -184,6 +203,38 @@ Keeps the full path of downloaded files from a Google Cloud Storage bucket. Requ
...
+
+### testTargetsAlwaysRun
+Always run - these tests are inserted at the beginning of every shard useful if you need to grant permissions or login before other tests run
+
+### otherFiles
+A list of device-path: file-path pairs that indicate the device paths to push files to the device before starting tests, and the paths of files to push. Device paths must be under absolute, whitelisted paths (${EXTERNAL_STORAGE}, or ${ANDROID_DATA}/local/tmp). Source file paths may be in the local filesystem or in Google Cloud Storage (gs://…).
+
+### networkProfile
+The name of the network traffic profile, for example LTE, HSPA, etc, which consists of a set of parameters to emulate network conditions when running the test (default: no network shaping; see available profiles listed by the `flank test network-profiles list` command). This feature only works on physical devices.
```
Comment (runningcode) [inline README.md:237]: nice, i like the examples and the command to fetch other supported values.

File: buildSrc/src/main/java/com/osacky/flank/gradle/FladleConfigImpl.kt [hunk 11]
```diff
@@ -24,5 +24,16 @@ data class FladleConfigImpl(
   override var resultsBucket: String? = null,
   override var keepFilePath: Boolean = false,
   override var resultsDir: String?,
-  override var additionalTestApks: Map<String, List<String>> = emptyMap()
+  override var additionalTestApks: Map<String, List<String>> = emptyMap(),
```
[the remark this answers is not in the document]
Comment (pawelpasterz) [reply buildSrc/src/main/java/com/osacky/flank/gradle/FladleConfigImpl.kt:27]: I was thinking about small refactor but decided to keep it separate from each other. I'd like to contribute more into fladle and next PR would be probably refactor of some pieces of code (if I may ofc)

[the remark this answers is not in the document]
Comment (runningcode) [reply buildSrc/src/main/java/com/osacky/flank/gradle/FladleConfigImpl.kt:27]: Yes, that would be lovely. and yes, its a good idea to keep these PRs separate.

**Comment by pawelpasterz** [comment]
Hey @runningcode thanks for review and comments. I implemented changes, please verify. Thanks!

File: README.md [hunk 12]
```diff
@@ -184,6 +203,44 @@ Keeps the full path of downloaded files from a Google Cloud Storage bucket. Requ
 ### resultsDir
 The name of a unique Google Cloud Storage object within the results bucket where raw test results will be stored. The default is a timestamp with a random suffix.
 
+### runTimeout
+The max time this test run can execute before it is cancelled. s (seconds), m (minutes), h (hours) suffixes are acceptable, mixes like 1h45m are currently not supported (default: unlimited).
```
Comment (runningcode) [inline README.md:207]: 👍

File: README.md [hunk 13]
```diff
@@ -184,6 +203,44 @@ Keeps the full path of downloaded files from a Google Cloud Storage bucket. Requ
...
+
+### ignoreFailedTests
+Always return successful task completion even when there are failed tests. Useful when parsing JUnit XML to determine failure. (default: false)
+
+### disableSharding
+Disables sharding. All tests will run on the same device. Useful for parameterized tests which do not support sharding. (default: false)
+
+### smartFlankDisableUpload
+Disables smart flank JUnit XML uploading. Useful for preventing timing data from being updated. (default: false)
+[What is Smart Flank?](https://github.com/Flank/flank/blob/master/docs/smart_flank.md)
```
Comment (runningcode) [inline README.md:221]: 👍

**Comment by pawelpasterz** [comment]
> Nice work! I'm going to wait until an updated version of Flank is released with these features before merging, is that okay? Otherwise people might look at the readme and try to use features that don't work with any released version.

Definitely, I will keep this PR updated if any new flag would appear in flank

**Comment by runningcode** [comment]
Hey! I was looking in to merging this, but when testing with Flank 8.1.0 i ran in to this issue which caused flank to hang:
```
com.fasterxml.jackson.module.kotlin.MissingKotlinParameterException: Instantiation of [simple type, class ftl.reports.xml.model.JUnitTestSuite] value failed for JSON property hostname due to missing (therefore NULL) value for creator parameter hostname which is a non-nullable type
 at [Source: (byte[])"<?xml version='1.0' encoding='UTF-8' ?>
<testsuites>
  <testsuite name="Nexus5-23-en-portrait" tests="1" failures="0" flakes="0" errors="0" skipped="0" time="2.594" timestamp="2020-04-23T20:40:34">
    <testcase name="seeView" classname="com.osacky.flank.gradle.sample.ExampleInstrumentedTest" time="2.594"/>
  </testsuite>
  <testsuite name="Pixel2-26-en-portrait" tests="1" failures="0" flakes="0" errors="0" skipped="0" time="2.712" timestamp="2020-04-23T20:40:14">
    <testcase name="seeView" cl"[truncated 109 bytes]; line: 5, column: 3] (through reference chain: ftl.reports.xml.model.JUnitTestResult["testsuite"]->java.util.ArrayList[0]->ftl.reports.xml.model.JUnitTestSuite["hostname"])
        at com.fasterxml.jackson.module.kotlin.KotlinValueInstantiator.createFromObjectWith(KotlinValueInstantiator.kt:107)
        at com.fasterxml.jackson.databind.deser.impl.PropertyBasedCreator.build(PropertyBasedCreator.java:198)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:488)
        at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1287)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:326)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:159)
        at com.fasterxml.jackson.dataformat.xml.deser.WrapperHandlingDeserializer.deserialize(WrapperHandlingDeserializer.java:114)
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:286)
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245)
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:27)
        at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:530)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:528)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:417)
        at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1287)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:326)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:159)
        at com.fasterxml.jackson.dataformat.xml.deser.WrapperHandlingDeserializer.deserialize(WrapperHandlingDeserializer.java:114)
        at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4202)
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3266)
        at ftl.reports.xml.JUnitXmlKt.parseAllSuitesXml(JUnitXml.kt:50)
        at ftl.reports.xml.JUnitXmlKt.parseAllSuitesXml(JUnitXml.kt:54)
        at ftl.gc.GcStorage.downloadJunitXml(GcStorage.kt:94)
        at ftl.args.ArgsHelper.calculateShards(ArgsHelper.kt:222)
        at ftl.args.AndroidTestShard.getTestShardChunks(AndroidTestShard.kt:26)
        at ftl.run.AndroidTestRunner$runTests$2.invokeSuspend(AndroidTestRunner.kt:38)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.ResumeModeKt.resumeUninterceptedMode(ResumeMode.kt:45)
        at kotlinx.coroutines.internal.ScopeCoroutine.afterCompletionInternal(Scopes.kt:32)
        at kotlinx.coroutines.JobSupport.completeStateFinalization(JobSupport.kt:310)
        at kotlinx.coroutines.JobSupport.tryFinalizeFinishingState(JobSupport.kt:236)
        at kotlinx.coroutines.JobSupport.tryMakeCompletingSlowPath(JobSupport.kt:849)
        at kotlinx.coroutines.JobSupport.tryMakeCompleting(JobSupport.kt:811)
        at kotlinx.coroutines.JobSupport.makeCompletingOnce$kotlinx_coroutines_core(JobSupport.kt:787)
        at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:111)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
        at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:241)
        at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:270)
        at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:79)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:54)
        at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:36)
        at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
        at ftl.cli.firebase.test.android.AndroidRunCommand.run(AndroidRunCommand.kt:49)
        at picocli.CommandLine.executeUserObject(CommandLine.java:1729)
        at picocli.CommandLine.access$900(CommandLine.java:145)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2101)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2068)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:1935)
        at picocli.CommandLine.execute(CommandLine.java:1864)
        at ftl.Main$Companion.main(Main.kt:41)
        at ftl.Main.main(Main.kt)
```

Are we adding a flag that doesn't work with the old version of flank?

File: buildSrc/src/main/java/com/osacky/flank/gradle/FladleConfigImpl.kt [hunk 14]
```diff
@@ -25,5 +25,16 @@ data class FladleConfigImpl(
   override var resultsBucket: String?,
   override var keepFilePath: Boolean,
   override var resultsDir: String?,
-  override var additionalTestApks: Map<String, List<String>>
+  override var additionalTestApks: Map<String, List<String>>,
+  override var runTimeout: String? = null,
```
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/FladleConfigImpl.kt:29]: btw, i updated the rest of this class, but you can remove the redundant default specification from this file!

File: buildSrc/src/main/java/com/osacky/flank/gradle/FladleConfig.kt [hunk 15]
```diff
@@ -55,4 +55,96 @@ interface FladleConfig {
...
+  var roboScript: String?
+
+  /**
+   * List of robo_directives that you can use to customize the behavior of Robo test.
+   * The type specifies the action type of the directive, which may take on values click, text or ignore.
+   * Each directive is list of String = [type, key, value]
+   * Each key should be the Android resource name of a target UI element and each value should be the text input for that element.
+   * Values are only permitted for text type elements, so no value should be specified for click and ignore type elements.
+   */
+  var roboDirectives: List<List<String>>
```
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/FladleConfig.kt:149]: ohhh. do we want this to be a `List<Triple>`?
Or is it better represented as `Map<String, List<String>>`?

File: buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt [hunk 16]
```diff
@@ -106,6 +128,44 @@ internal class YamlWriter {
...
+    }
+
+    config.roboScript?.let {
+      appendln("  robo-script: $it")
+    }
+
+    if (config.roboDirectives.isNotEmpty()) {
+      appendln("  robo-directives:")
+      config.roboDirectives.forEach {
+        val value = it.getOrNull(2).let { stringValue -> if (stringValue.isNullOrBlank()) "\"\"" else stringValue }
```
Comment (runningcode) [inline buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt:165]: if `it.getOrNull(2)` returns null, does it crash on `.let`? should it be `?.let`?

**Comment by runningcode** [comment]
Looking good. I'll test this tomorrow now that I figured out the issue that was causing the crash for me!

Also, I won't block this PR since it already is quite large, but I just wanted to also give you a heads up: From now on, I would like new extensions in fladle to use [Gradle's lazy property API](https://docs.gradle.org/current/userguide/lazy_configuration.html). This will allow for lazy evaluation and faster configuration times.

I started doing some migration here: https://github.com/runningcode/fladle/commit/f92bd12927db99e6cfd10aa32883df465183ef74#diff-670e340c5b646cd261359b8466643022R40

Anyways, don't worry about it. I'll do the migration for you once I merge this PR.

Let's figure out that robo directives api first!

File: buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt (hunk 16 above)
Comment (pawelpasterz) [reply buildSrc/src/main/java/com/osacky/flank/gradle/YamlWriter.kt:165]: Interesting fact it worked! But I think you are right.  I'll change it to be more robust

File: buildSrc/src/main/java/com/osacky/flank/gradle/FladleConfig.kt (hunk 15 above)
Comment (pawelpasterz) [reply buildSrc/src/main/java/com/osacky/flank/gradle/FladleConfig.kt:149]: I was thinking about that but I wanted to allow user to skip 3rd parameter.
```
roboDirectives = [
      ["click", "button1", ""],
      ["ignore", "button2"],  <-- 3rd skipped
      ["text", "field1", "my text"],
    ]
```
`Triple` will force user to always provide three strings. But I am open to suggestions

**Comment by bootstraponline** [comment]
@pawelpasterz now that flank is released, let's double check all the flags are supported.

I think `additional-apks` might be missing? https://github.com/Flank/flank/pull/695

**Comment by runningcode** [comment]
Hey, I just tested this out. I'm going to merge to master and release a snapshot! Thanks!

**Comment by andrey-bolduzev** [comment]
Hey @runningcode, have you had a chance to push these exciting changes to a snapshot release? I don't think `0.9.2-SNAPSHOT` includes them. Can't wait to try! Thank you.

**Comment by runningcode** [comment]
There may be an older `0.9.2-SNAPSHOT` being pulled from the Gradle Plugin Repository overriding the one from the sonatype snaphots repo. Can you put the sonatype snapshots repo above the gradle plugin repository? Or add the following beneath the snapshots repository declaration?
```
mavenContent {
  snapshotsOnly()
}
```

More info here: https://docs.gradle.org/current/userguide/declaring_repositories.html

**Comment by runningcode** [comment]
Hey, I've deleted the `0.9.2-SNAPSHOT` from the Gradle Plugins portal. Please re-run using `--refresh-dependencies` and it should work. Let me know if you run in to any trouble.

**Comment by andrey-bolduzev** [comment]
> Hey, I've deleted the `0.9.2-SNAPSHOT` from the Gradle Plugins portal. Please re-run using `--refresh-dependencies` and it should work. Let me know if you run in to any trouble.

To be honest I'm having trouble figuring out the correct coordinates for the sonatype snapshots repository. I _think_ it's `https://oss.sonatype.org/service/local/repositories/snapshots/content/` which I added to the `settings.gradle.kts`'s `pluginManagement`, but then the plugin with this id cannot be found there, so I suspect there's some extra mapping needed to be done. Would it be possible to upload the latest snapshot to the main gradle plugin repository? Or I could wait until the next regular release.

**Comment by runningcode** [comment]
I believe the URL is this: `https://oss.sonatype.org/content/repositories/snapshots`

**Pull request merged**